frame/election-provider-multi-phase/src/signed.rs

```diff
@@ -0,0 +1,662 @@
+// This file is part of Substrate.
+
+// Copyright (C) 2020 Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! The signed phase implementation.
+
+use crate::{
+	CompactOf, Config, ElectionCompute, Pallet, RawSolution, ReadySolution, SolutionOrSnapshotSize,
+	Weight, WeightInfo, QueuedSolution, SignedSubmissions,
+};
+use codec::{Encode, Decode, HasCompact};
+use frame_support::traits::{Currency, Get, OnUnbalanced, ReservableCurrency};
+use sp_arithmetic::traits::SaturatedConversion;
+use sp_npos_elections::{is_score_better, CompactSolution};
+use sp_runtime::{Perbill, RuntimeDebug, traits::Zero};
+use sp_std::{cmp::Ordering, vec::Vec};
+
+/// A raw, unchecked signed submission.
+///
+/// This is just a wrapper around [`RawSolution`] and some additional info.
+#[derive(PartialEq, Eq, Clone, Encode, Decode, RuntimeDebug, Default)]
+pub struct SignedSubmission<AccountId, Balance: HasCompact, CompactSolution> {
+	/// Who submitted this solution.
+	pub who: AccountId,
+	/// The deposit reserved for storing this solution.
+	pub deposit: Balance,
+	/// The reward that should be given to this solution, if chosen the as the final one.
+	pub reward: Balance,
+	/// The raw solution itself.
+	pub solution: RawSolution<CompactSolution>,
+}
+
+pub type BalanceOf<T> =
+	<<T as Config>::Currency as Currency<<T as frame_system::Config>::AccountId>>::Balance;
+pub type PositiveImbalanceOf<T> = <<T as Config>::Currency as Currency<
+	<T as frame_system::Config>::AccountId,
+>>::PositiveImbalance;
+pub type NegativeImbalanceOf<T> = <<T as Config>::Currency as Currency<
+	<T as frame_system::Config>::AccountId,
+>>::NegativeImbalance;
+
+impl<T: Config> Pallet<T> {
+	/// Finish the signed phase. Process the signed submissions from best to worse until a valid one
+	/// is found, rewarding the best one and slashing the invalid ones along the way.
+	///
+	/// Returns true if we have a good solution in the signed phase.
+	///
+	/// This drains the [`SignedSubmissions`], potentially storing the best valid one in
+	/// [`QueuedSolution`].
+	pub fn finalize_signed_phase() -> (bool, Weight) {
+		let mut all_submissions: Vec<SignedSubmission<_, _, _>> = <SignedSubmissions<T>>::take();
```

Comment:
As suggested above, using a map here will be more efficient than a single huge storage item, and seems it will work just fine.

Reply:
The hard part IMO is the ordering. The correctness of the algorithm depends on the order in which we process the signed submissions, which means that if we use a storage map instead, we'll need to add some kind of additional storage item to keep track of the order in which the values should be processed.